internal/controller/apiextensions/composite/composed.go

```diff
@@ -87,22 +311,36 @@ func NewAPIDryRunRenderer(c client.Client) *APIDryRunRenderer {
 // and template. The rendered resource may be submitted to an API server via a
 // dry run create in order to name and validate it.
 func (r *APIDryRunRenderer) Render(ctx context.Context, cp resource.Composite, cd resource.Composed, t v1.ComposedTemplate) error {
-	// Any existing name will be overwritten when we unmarshal the template. We
-	// store it here so that we can reset it after unmarshalling.
+	kind := cd.GetObjectKind().GroupVersionKind().Kind
 	name := cd.GetName()
 	namespace := cd.GetNamespace()
+
 	if err := json.Unmarshal(t.Base.Raw, cd); err != nil {
 		return errors.Wrap(err, errUnmarshal)
 	}
+
+	// We think this composed resource exists, but when we rendered its template
+	// its kind changed. This shouldn't happen. Either someone changed the kind
+	// in the template or we're trying to use the wrong template (e.g. because
+	// the order of an array of anonymous templates changed).
+	if kind != "" && cd.GetObjectKind().GroupVersionKind().Kind != kind {
+		return errors.New(errKindChanged)
+	}
```

Comment:
This is intended to be an optimistic safeguard. Without it there's a good chance that reordering an array of anonymous Composition resource templates will cause all XRs that use said Composition to orphan and recreate a subset of their existing composed resources. The 'orphaned' resources won't be orphaned in the controller reference sense, but will be gone from the resource references array.

With this safeguard in place we'll return an error in the event that the reordering would change the kind of a composed resource; i.e. if the template at index was used to create an `RDSInstance`, but now represents an `IAMRole`. This won't help if two `RDSInstance` templates are switched, but it's better than nothing.